cmd/nodecmd/upgrade.go

```diff
@@ -0,0 +1,224 @@
+	"golang.org/x/exp/slices"
+)
+
+func newUpgradeCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "upgrade",
+		Short: "(ALPHA Warning) Update avalanchego or VM version for all node in a cluster",
+		Long: `(ALPHA Warning) This command is currently in experimental mode.
+
+The node update command suite provides a collection of commands for nodes to update
+their avalanchego or VM version.
+
+You can check the status after upgrade by calling avalanche node status`,
+		SilenceUsage: true,
+		Args:         cobra.ExactArgs(1),
+		RunE:         upgrade,
+	}
+
+	return cmd
+}
+
+func upgrade(_ *cobra.Command, args []string) error {
+	clusterName := args[0]
+	if err := setupAnsible(clusterName); err != nil {
+		return err
+	}
+	toUpgradeNodes, err := getNodesToBeUpgraded(clusterName)
+	if err != nil {
+		return err
+	}
+	for node, toUpgrade := range toUpgradeNodes {
+		// toUpgradeNodes either has value of avalancheGo or subnetEvm or avalancheGo,subnetEvm
+		toUpgradeItems := strings.Split(toUpgrade, ",")
+		for _, toUpGradeItem := range toUpgradeItems {
+			// toUpGradeItem either has value format of avalancheGo=<versionNum> or subnetEvm=<versionNum>
+			upgradeInfo := strings.Split(toUpGradeItem, "=")
+			if strings.Contains(toUpGradeItem, constants.AvalancheGoRepoName) {
+				if err = upgradeAvalancheGo(clusterName, node, upgradeInfo[1]); err != nil {
+					return err
+				}
+			} else if strings.Contains(toUpGradeItem, constants.SubnetEVMRepoName) {
+				// subnetEVM version has value format of n8Anw9kErmgk7KHviddYtecCmziLZTphDwfL1V2DfnFjWZXbE:<versionNum>
+				subnetEVMVersionInfo := strings.Split(upgradeInfo[1], ":")
+				subnetEMVersionToUpgradeTo := subnetEVMVersionInfo[1]
+				subnetEMVersionToUpgradeToWoPrefix := strings.TrimPrefix(subnetEMVersionToUpgradeTo, "v")
+				subnetEVMReleaseURL := fmt.Sprintf("https://github.com/ava-labs/subnet-evm/releases/download/%s/subnet-evm_%s_linux_amd64.tar.gz", subnetEMVersionToUpgradeTo, subnetEMVersionToUpgradeToWoPrefix)
+				subnetEVMArchive := fmt.Sprintf("subnet-evm_%s_linux_amd64.tar.gz", subnetEMVersionToUpgradeToWoPrefix)
+				subnetEVMBinaryPath := fmt.Sprintf("/home/ubuntu/.avalanchego/plugins/%s", subnetEVMVersionInfo[0])
+				if err = upgradeSubnetEVM(clusterName, subnetEVMReleaseURL, subnetEVMArchive, subnetEVMBinaryPath, node, subnetEMVersionToUpgradeTo); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// getNodesToBeUpgraded gets the node versions of all nodes in cluster clusterName and checks which
+// nodes needs to have Avalanche Go & SubnetEVM upgraded. It first checks the subnet EVM version -
+// it will install the newest subnet EVM version and install the latest avalanche Go that is still compatible with the Subnet EVM version
+// if the node is not tracking any subnet, it will just install latestAvagoVersion
+// sample return result map[aws_node_i-006e0fbf1df7e2198:avalanchego=v1.10.13,subnet-evm=n8Anw9kErmgk7KHuuxRag42nR4xhhBeePH2MkwSxfGNPiaCJs:v0.5.7
```

Comment:
just return a map of node ids to some struct containing all the upgrade info

Reply:
or if you like so, a list of structs where the node id is.also a field of the struct

Reply:
or if you like so, a list of structs where the node id is.also a field of the struct

Reply:
new idea: enable parallelization. have a list of nodes associated to each upgrade requirement. also improve the subnet-evm command to upgrade a list of subnet-evm binaries in one command call (probably give to ansible to vm ids not the full vm path)